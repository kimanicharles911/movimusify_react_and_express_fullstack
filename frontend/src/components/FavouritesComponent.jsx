import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart } from '@fortawesome/free-solid-svg-icons';

const FavouritesComponent = ({favouritesProp, setFavouritesProp}) => {

  const unlikeFunc = (dataIndex) => {
    const newFavouritesArr = favouritesProp.filter(favourite => favourite !== favouritesProp[dataIndex]);
    setFavouritesProp(newFavouritesArr);
  };

  return(
    <main className="container fluid text-center">
      <section className="row justify-content-md-center mt-3 mb-3 gy-3">
        {favouritesProp.length > 0 ?
          favouritesProp.map((dataPoint, index) => {
            return (
              <div className="col-sm-6" key={index}>
                <div className="card mb-3" style={{ maxWidth: "540px" }}>
                  <div className="row g-0">
                    <div className="col-md-4">
                      <img src={dataPoint.artworkUrl100} alt="media art image" className="img-fluid rounded-start" />
                    </div>
                    <div className="col-md-8">
                      <div className="card-body">
                        <h5 className="card-title"><strong>{dataPoint.trackName}</strong> by {dataPoint.artistName}</h5>
                        <p className="card-text">Cost: ${dataPoint.trackPrice} Dollars</p>
                        <span>&nbsp;|&nbsp;</span>
                        <p className="card-text">Genre: {dataPoint.primaryGenreName}</p>
                        <br/>
                        <p className="card-text"><small className="text-muted">Kind: {dataPoint.kind}</small></p>
                        <span>&nbsp;|&nbsp;</span>
                        <p className="card-text"><small className="text-muted small-like-tag" onClick={() => unlikeFunc(index) }>{dataPoint.likeStatus ?  'Unlike' : 'Like'}<a type="button" className="nav-link like-icons"><FontAwesomeIcon icon={faHeart} className="font-awesome-icon-component" /></a></small></p>
                        <p className="card-text">Description: {dataPoint.description ? dataPoint.description : null } {dataPoint.description === undefined && dataPoint.shortDescription ? dataPoint.shortDescription : "Not available." } </p>
                        <br />
                        <p className="card-text">Long Description: {dataPoint.longDescription ? dataPoint.longDescription : "Not available." }</p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            )
          })
          : "You have not liked any media so far. Please start by liking and come back here to view them as your favourites."
        }
      </section>
    </main>
  )
};

export default FavouritesComponent;

/* 
  * I first imported the FontAwesomeIcon component from the @fortawesome library
  * I then imported the faHeart component from @fortawesome library
  * I then created the FavouritesComponent and destructured the favouritesProp and setFavouritesProp.
  * I then wrote the unlikeFunc which receives the array index of a favourite item as it's parameter. It then removes the unliked media from the favouritesProp which trigger changes in the UI.
  * I then wrote the JSX of the FavouritesComponent.
  * If one or more favourites exist a map JS function is used to render the JSX for each of the favourites and dynamically sets the data of every favourite.
  * If there are no favourites the user is shown the appropriate message.
  * I then export the FavouritesComponent.
*/