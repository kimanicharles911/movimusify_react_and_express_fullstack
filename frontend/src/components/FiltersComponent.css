#filters-section > .btn{
  color: #6c757d;
  background-color: #fff;
  box-shadow: none;
}
#btn-group > label{
  border-radius: 0.25rem;
}
/* 
  * I changed the color and background-color of the filter buttons and gave them a shadow.
  * I gave the button labels a radius
*/