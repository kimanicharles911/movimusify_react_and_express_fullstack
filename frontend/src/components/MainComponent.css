#search-btn{
  border-radius: 0.25rem;
  color: #fff;
  background-color: #008075;
  border-color: #008075;
}
#filters-section > .btn{
  color: #6c757d;
  background-color: #fff;
  box-shadow: none;
}
#filters-section > .trial-btn{
  color: #fff;
  background-color: #008075;
  box-shadow: none;
}
#btn-group > label{
  border-radius: 0.25rem;
}
.card-body > .card-text, .card-body > .card-text > .text-muted > .like-icons{
  display: inline-block;
}
.like-icons{
  color: inherit;
}
.font-awesome-icon-component{
  color: #6c757d;
}
.like-icons > .fas{
  color: #008075;
}
.like-icons:hover{
  color: inherit;
}
#spinner-border{
  border-top-color: #008075;
  border-bottom-color: #008075;
  border-left-color: #008075;
}
.small-like-tag{
  cursor: pointer;
}