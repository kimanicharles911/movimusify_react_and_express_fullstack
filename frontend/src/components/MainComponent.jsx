import {useState} from 'react';
import {SearchComponent, FiltersComponent, ResultsComponent} from './index.js';

const MainComponent = ({favouritesProp, setFavouritesProp}) => {

  const [searchStore, setSearchStore] = useState({
    typedInput: '',
    clickedFilter: '',
  });
  const [apiData, setApiData] = useState([]);
  const [spin, setSpin] = useState(false);

  const modifyTypedInput = () => {
    const arrFromInput = [];
    searchStore.typedInput.split('').map(arrItem => {
      if(arrItem === ' ') arrItem = '+';
      arrFromInput.push(arrItem);
    })
    return arrFromInput.join('');
  };

  const searchFunc = async(filterBtn = false) => {
    if(searchStore.typedInput){
      setSpin(true);
      let dataArr;
      try{
        let res;
        if(filterBtn === false){
          res = await fetch('/search/?term='+ modifyTypedInput() + searchStore.clickedFilter);
        }else{
          res = await fetch('/search/?term='+ modifyTypedInput() + '&media=' + filterBtn);
        }
        let data = await res.json();
        dataArr = data.results;
      }catch(err){
        console.error('err ',err);
        dataArr = [];
      }
      setApiData(dataArr);
      setSpin(false);
    }
  };

  const likeUnlikeFunc = (dataIndex) => {
    if(favouritesProp.length > 0){
      if(!favouritesProp.includes(apiData[dataIndex])){
        setFavouritesProp((prevState) => {
          return[
            ...prevState,
            apiData[dataIndex]
          ]
        })

        apiData[dataIndex].likeStatus = true;
      }else{
        const newFavouritesArr = favouritesProp.filter(favourite => favourite !== apiData[dataIndex]);
        setFavouritesProp(newFavouritesArr);
        apiData[dataIndex].likeStatus = false;
      }
    }else{
      setFavouritesProp((prevState) => {
        return[
          ...prevState,
          apiData[dataIndex]
        ]
      })

      apiData[dataIndex].likeStatus = true;
    }
  };

  return (
    <main className="container fluid text-center">
      <SearchComponent searchStoreProp={searchStore} setSearchStoreProp={setSearchStore}spinProp={spin} setSpinProp={setSpin} searchFuncProp={searchFunc}/>
      <FiltersComponent setSearchStoreProp={setSearchStore} searchFuncProp={searchFunc}/>
      <hr />
      <ResultsComponent apiDataProp={apiData} setApiDataProp={setApiData} likeUnlikeFuncProp={likeUnlikeFunc}/>
    </main >
  )
};
export default MainComponent;
/*
  * I wrote the JSX for the main part of the application.
  * The main function in the application is a map function that creates bootstrap cards with the name and description of each repository it loops through.
  * The other main JSX items are the Delete, Edit and View Buttons which are used to open their respective modals. They also send the repository name to the repoNameReceiverFunc function so as to enable identification of the project the user is interacting with.
  * I then wrapped the other three main components i.e ViewModalComponent, EditModalComponent and DeleteModalComponent.
  * In each of the above components I passed the props needed for the modal to operate as expected.
*/