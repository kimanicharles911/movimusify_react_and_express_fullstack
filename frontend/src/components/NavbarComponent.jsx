import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {Navbar, Container, Nav, Row} from 'react-bootstrap';
import './NavbarComponent.css';
import {faHeart} from '@fortawesome/free-solid-svg-icons';
import {faGithub} from '@fortawesome/free-brands-svg-icons';

/* 
  * I imported the Navbar, Container, Nav, Row components from react-bootstrap.
  * I imported the css style file i.e NavbarComponent.css.
  * I imported the FontAwesomeIcon component and faPlus icon which are all products of the font-awesome library.
  * I imported the CreateModalComponent component from the components folder.
*/

const NavbarComponent = () => {

  return (
    <>
      <Navbar className="nav-class" expand="sm">
        <Container fluid>
          {/* Application Name */}
          <Navbar.Brand href="#" id="brand-id">MOVIMUSIFY</Navbar.Brand>
          <Navbar.Toggle aria-controls="basic-navbar-nav" id="navbar-toggler" />
          <Navbar.Collapse id="basic-navbar-nav">
            <Container>
              <Row>
                <Nav className="me-auto">
                  <ul className="navbar-nav me-auto mb-2 mb-sm-0">           
                  </ul>
                  <ul className="d-flex navbar-nav me-2 mb-2 mb-sm-0">
                    {/* Favourites Link */}
                    <Nav.Link href="https://github.com/kimanicharles911/emmethub_github_projects_react_and_express_fullstack" target="_blank" id="nav-link-class"><FontAwesomeIcon icon={faHeart} className="font-awesome-icons" id="heart-icon"/>&nbsp;Favourites</Nav.Link>
                    
                    {/* Github Repository Link Button */}
                    <Nav.Link href="https://github.com/kimanicharles911/movimusify_react_and_express_fullstack" target="_blank" id="nav-link-class"><FontAwesomeIcon icon={faGithub} className="font-awesome-icons" id="github-icon"/>&nbsp;GitHub</Nav.Link>
                  </ul>
                </Nav>
              </Row>
            </Container>
          </Navbar.Collapse>
        </Container>
      </Navbar>
    </>
  )
};

export default NavbarComponent;

/* 
  * I created a Navbar component.
  * I then passed then destructured the renderAgentProp and setRenderAgentProp received from the App component.
  * I then created the navbar JSX that heavily relies on bootstrap and react-bootstrap libraries.
  * The first major JSX item is the Create Project Button that is used to open the Create Project Modal.
  * The only wrapped component is the CreateModalComponent. I pass the renderAgentProp and setRenderAgentProp to it also.
  * The second major JSX item is the Github Repository Link Button that redirects the user to this Application's source code repository.
*/