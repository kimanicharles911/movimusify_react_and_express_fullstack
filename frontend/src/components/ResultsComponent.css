.card-body > .card-text, .card-body > .card-text > .text-muted > .like-icons{
  display: inline-block;
}
.like-icons{
  color: inherit;
}
.font-awesome-icon-component{
  color: #6c757d;
}
.like-icons > .fas{
  color: #008075;
}
.like-icons:hover{
  color: inherit;
}
#spinner-border{
  border-top-color: #008075;
  border-bottom-color: #008075;
  border-left-color: #008075;
}
.small-like-tag{
  cursor: pointer;
}