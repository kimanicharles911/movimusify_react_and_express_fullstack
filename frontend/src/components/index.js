export {default as NavbarComponent} from './NavbarComponent.jsx';
export {default as MainComponent} from './MainComponent.jsx';
export {default as FavouritesComponent} from './FavouritesComponent.jsx';
export {default as SearchComponent} from './SearchComponent.jsx';