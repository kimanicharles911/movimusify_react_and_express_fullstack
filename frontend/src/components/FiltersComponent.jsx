import './FiltersComponent.css';

const FiltersComponent = ({setSearchStoreProp, searchFuncProp}) => {

  const filterBtnStyling = "btn btn-outline-secondary btn-sm mb-1 mx-1";

  const filterBtnClickHandler = (filterBtn) => {
    setSearchStoreProp((prevState) => {
      return{
        ...prevState,
        clickedFilter: `&media=${filterBtn}`
      }
    })
    searchFuncProp(filterBtn);
  };

  return(
    <section id="filters-section">
      <p>Click to Apply Filters jack johnson</p>
      <div className="btn-group flex-wrap" id="btn-group" role="group">
        <input type="radio" className="btn-check" name="btnradio" id="btnradioMovie"/>
        <label className={filterBtnStyling}  htmlFor="btnradioMovie" onClick={() => {filterBtnClickHandler("movie")}}>movie</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioMusic"/>
        <label className={filterBtnStyling}  htmlFor="btnradioMusic" onClick={() => {filterBtnClickHandler("music")}}>music</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioMusicVideo"/>
        <label className={filterBtnStyling}  htmlFor="btnradioMusicVideo" onClick={() => {filterBtnClickHandler("musicVideo")}}>music video</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioPodcast"/>
        <label className={filterBtnStyling}  htmlFor="btnradioPodcast" onClick={() => {filterBtnClickHandler("podcast")}}>podcast</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioAudiobook"/>
        <label className={filterBtnStyling}  htmlFor="btnradioAudiobook" onClick={() => {filterBtnClickHandler("audiobook")}}>audiobook</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioShortFilm"/>
        <label className={filterBtnStyling}  htmlFor="btnradioShortFilm" onClick={() => {filterBtnClickHandler("shortFilm")}}>short film</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioTvShow"/>
        <label className={filterBtnStyling}  htmlFor="btnradioTvShow" onClick={() => {filterBtnClickHandler("tvShow")}}>TV show</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioSoftware"/>
        <label className={filterBtnStyling}  htmlFor="btnradioSoftware" onClick={() => {filterBtnClickHandler("software")}}>software</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioEbook"/>
        <label className={filterBtnStyling}  htmlFor="btnradioEbook" onClick={() => {filterBtnClickHandler("ebook")}}>ebook</label>
        <input type="radio" className="btn-check" name="btnradio" id="btnradioAll"/>
        <label className={filterBtnStyling}  htmlFor="btnradioAll" onClick={() => {filterBtnClickHandler("all")}}>All</label>
      </div>
    </section>
  )
}
export default FiltersComponent;