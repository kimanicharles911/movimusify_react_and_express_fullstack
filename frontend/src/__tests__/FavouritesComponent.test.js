import { expectCt } from 'helmet';
import renderer from 'react-test-renderer';
import {FavouritesComponent} from '../components';

it('FavouritesComponent renders correctly', () => {
  const favourites = [
    {
      id: 1,
      title: "delectus aut autem",
    },
    {
      id: 2,
      title: "quis ut nam facilis et officia qui",
    },
    {
      id: 3,
      title: "fugiat veniam minus",
    }
  ];
  const setFavourites = () => {
    return favourites;
  };
  const tree = renderer.create(<FavouritesComponent favouritesProp={favourites} setFavouritesProp={setFavourites}/>).toJSON();
  expect(tree).toMatchSnapshot();
})