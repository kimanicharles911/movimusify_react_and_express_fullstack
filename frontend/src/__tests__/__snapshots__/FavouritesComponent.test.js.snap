// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavouritesComponent renders correctly 1`] = `
<main
  className="container fluid text-center"
>
  <section
    className="row justify-content-md-center mt-3 mb-3 gy-3"
  >
    <div
      className="col-sm-6"
    >
      <div
        className="card mb-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="media art image"
              className="img-fluid rounded-start"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                <strong />
                 by 
              </h5>
              <p
                className="card-text"
              >
                Cost: $
                 Dollars
              </p>
              <span>
                 | 
              </span>
              <p
                className="card-text"
              >
                Genre: 
              </p>
              <br />
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Kind: 
                </small>
              </p>
              <span>
                 | 
              </span>
              <p
                className="card-text"
              >
                <small
                  className="text-muted small-like-tag"
                  onClick={[Function]}
                >
                  Like
                  <a
                    className="nav-link like-icons"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-heart fa-w-16 font-awesome-icon-component"
                      data-icon="heart"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={Object {}}
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </a>
                </small>
              </p>
              <p
                className="card-text"
              >
                Description: 
                 
                Not available.
                 
              </p>
              <br />
              <p
                className="card-text"
              >
                Long Description: 
                Not available.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-sm-6"
    >
      <div
        className="card mb-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="media art image"
              className="img-fluid rounded-start"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                <strong />
                 by 
              </h5>
              <p
                className="card-text"
              >
                Cost: $
                 Dollars
              </p>
              <span>
                 | 
              </span>
              <p
                className="card-text"
              >
                Genre: 
              </p>
              <br />
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Kind: 
                </small>
              </p>
              <span>
                 | 
              </span>
              <p
                className="card-text"
              >
                <small
                  className="text-muted small-like-tag"
                  onClick={[Function]}
                >
                  Like
                  <a
                    className="nav-link like-icons"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-heart fa-w-16 font-awesome-icon-component"
                      data-icon="heart"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={Object {}}
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </a>
                </small>
              </p>
              <p
                className="card-text"
              >
                Description: 
                 
                Not available.
                 
              </p>
              <br />
              <p
                className="card-text"
              >
                Long Description: 
                Not available.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-sm-6"
    >
      <div
        className="card mb-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="media art image"
              className="img-fluid rounded-start"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                <strong />
                 by 
              </h5>
              <p
                className="card-text"
              >
                Cost: $
                 Dollars
              </p>
              <span>
                 | 
              </span>
              <p
                className="card-text"
              >
                Genre: 
              </p>
              <br />
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Kind: 
                </small>
              </p>
              <span>
                 | 
              </span>
              <p
                className="card-text"
              >
                <small
                  className="text-muted small-like-tag"
                  onClick={[Function]}
                >
                  Like
                  <a
                    className="nav-link like-icons"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-heart fa-w-16 font-awesome-icon-component"
                      data-icon="heart"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={Object {}}
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
                        fill="currentColor"
                        style={Object {}}
                      />
                    </svg>
                  </a>
                </small>
              </p>
              <p
                className="card-text"
              >
                Description: 
                 
                Not available.
                 
              </p>
              <br />
              <p
                className="card-text"
              >
                Long Description: 
                Not available.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</main>
`;
